---
title: "Counting RT mutations from illumina sequencing data."
author: "atb abelew@gmail.com"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
  rmdformats::readthedown:
    code_download: true
    code_folding: show
    df_print: paged
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    width: 300
    keep_md: false
    mode: selfcontained
    toc_float: true
  BiocStyle::html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: tango
    keep_md: false
    mode: selfcontained
    toc_float: true
---

<style type="text/css">
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
 font-size: 16px
}
</style>

```{r options, include=FALSE}
library("hpgltools")
tt <- devtools::load_all("~/hpgltools")
knitr::opts_knit$set(width=120,
                     progress=TRUE,
                     verbose=TRUE,
                     echo=TRUE)
knitr::opts_chunk$set(error=TRUE,
                      dpi=96)
old_options <- options(digits=4,
                       stringsAsFactors=FALSE,
                       knitr.duplicate.label="allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size=10))
rundate <- format(Sys.Date(), format="%Y%m%d")
previous_file <- "tag.Rmd"
ver <- "202106"

##tmp <- sm(loadme(filename=paste0(gsub(pattern="\\.Rmd", replace="", x=previous_file), "-v", ver, ".rda.xz")))
rmd_file <- "error_quant_202106.Rmd"
```

# Calculating error rates.

I wrote the function 'create_matrices()' to collect mutation counts.  At least
in theory the results from it should be able to address most/any question
regarding the counts of mutations observed in the data.

For the moment at least, the only real change here is that we are
getting ready to submit this for publication and we added a filter for
hyper-mutated reads.  I think that has been successfully implemented
and so will run a fresh invocation of this.

# An offset

Previous experiments using similar constructs numbered the positions
of mutations slightly differently.  Here is the explanation from
Dr. DeStefano:

Number 6 in the (previous?) paper corresponds to number 22 in our
analysis.  Re-number based on that would work great. 

So... we should have an offset of 22 - x = 6, or -16.

As a result of this, I will remove 16 from the minimum/maximum filters.

# Recent samples

Each of these blocks of code produce extremely verbose outputs.  The
provide tables of how many mutations were observed per read, how many
reads are filtered at each step in the process, then information about
the categorization of mutations in the filtered data.

```{r 01params}
devtools::load_all("Rerrrt")
ident_column <- "identtable"
mut_column <- "mutationtable"
min_reads <- 3
min_tags <- 3
min_sequencer <- 6
## I made this the default argument for the function, so will not change my calls.
position_offset <- -16
min_position <- 6
max_position <- 169
max_mutations_per_read <- NULL
prune_n <- TRUE
verbose <- TRUE
plot_order <- c("dna_control", "dna_low", "dna_high", "rna_control", "rna_low", "rna_high")
type <- "recent"
sample_sheet <- glue("sample_sheets/{type}_samples_2020.xlsx")
excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")
```

## 3 min reads, 3 min tags, 6 sequencer, No MPR, 22<=pos<=185

For this first iteration we will use the following criteria:

1.  There must be at least 3 reads to consider an tag.
2.  There must be at least 3 tags to consider a mutation.
3.  In order for a mutation to be considered sequencer based, it must
    be present in exactly 1 of at least 6 reads in a tag group.
4.  We will ignore anything before position 22.
5.  We will ignore anything after position 185.
6.  We will not have a maximum number of mutations per read.
7.  We will prune any Ns in the reads.
8.  This analysis will include the 'recent' samples, which includes
    groups of controls for both DNA and RNA as well samples where high
    and low concentrations of Magnesium was included.

```{r 01triples}
recent_3mr_3mi_6s_0mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)
```

I am now collecting a histogram of the reads/tag before and I guess after processing:

```{r histogram1}
recent_3mr_3mi_6s_0mpr[["preprocessed_hist"]]
```

```{r 02recent_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(recent_3mr_3mi_6s_0mpr[["matrices"]])) {
  table_name <- names(recent_3mr_3mi_6s_0mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(recent_3mr_3mi_6s_0mpr[["matrices"]][table_name]))
  if (!is.null(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(recent_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])) {
  table_name <- names(recent_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(recent_3mr_3mi_6s_0mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(recent_3mr_3mi_6s_0mpr[["matrices_counts"]])) {
  table_name <- names(recent_3mr_3mi_6s_0mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(recent_3mr_3mi_6s_0mpr[["matrices_counts"]][table_name]))
  if (!is.null(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(recent_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(recent_3mr_3mi_6s_0mpr)
```

# All samples

## 3 min reads, 3 min tags, 6 sequencer, No MPR, 22<=pos<=185

```{r 04all_triples}
## Repeat the same parameters using all samples
sample_sheet <- "sample_sheets/all_samples_202101.xlsx"
type <- "all"
excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")
all_3mr_3mi_6s_0mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

all_3mr_3mi_6s_0mpr[["preprocessed_hist"]]
```

```{r 05all_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(all_3mr_3mi_6s_0mpr[["matrices"]])) {
  table_name <- names(all_3mr_3mi_6s_0mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(all_3mr_3mi_6s_0mpr[["matrices"]][table_name]))
  if (!is.null(all_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(all_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(all_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])) {
  table_name <- names(all_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(all_3mr_3mi_6s_0mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(all_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(all_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(all_3mr_3mi_6s_0mpr[["matrices_counts"]])) {
  table_name <- names(all_3mr_3mi_6s_0mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(all_3mr_3mi_6s_0mpr[["matrices_counts"]][table_name]))
  if (!is.null(all_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(all_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(all_3mr_3mi_6s_0mpr)
```

# RNA samples

## 3 min reads, 3 min tags, 6 sequencer, No MPR, 22<=pos<=185

```{r 06only_recent_triples}
## Repeat with only the recent RNA samples
sample_sheet <- "sample_sheets/rna_samples_202101.xlsx"
type <- "rna"

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_3mi_6s_0mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_3mi_6s_0mpr[["processed_hist"]]
```

```{r 08rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_3mi_6s_0mpr[["matrices"]])) {
  table_name <- names(rna_3mr_3mi_6s_0mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_0mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_3mi_6s_0mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_0mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_3mi_6s_0mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_3mi_6s_0mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_0mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_0mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_3mi_6s_0mpr)
```

# RNA samples

## 3 min reads, 3 min tags, 6 sequencer, 10 MPR, 22<=pos<=185

```{r 10rna_ten_mpr}
max_mutations_per_read <- 10

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_3mi_6s_10mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_3mi_6s_10mpr[["processed_hist"]]
```

```{r 11rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_3mi_6s_10mpr[["matrices"]])) {
  table_name <- names(rna_3mr_3mi_6s_10mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_10mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_3mi_6s_10mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_3mi_6s_10mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_10mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_3mi_6s_10mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_3mi_6s_10mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_10mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_10mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_3mi_6s_10mpr)
```

# RNA samples

## 3 min reads, 3 min tags, 6 sequencer, 5 MPR, 22<=pos<=185

```{r 12rna_triples_fivempr}
max_mutations_per_read <- 5

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_3mi_6s_5mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_3mi_6s_5mpr[["processed_hist"]]
```

```{r 13rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_3mi_6s_5mpr[["matrices"]])) {
  table_name <- names(rna_3mr_3mi_6s_5mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_5mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_3mi_6s_5mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_3mi_6s_5mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_5mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_3mi_6s_5mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_3mi_6s_5mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_3mi_6s_5mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_3mi_6s_5mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_3mi_6s_5mpr)
```

# RNA samples

## 3 min reads, 5 min tags, 6 sequencer, No MPR, 22<=pos<=185

```{r 15quints}
min_tags <- 5
max_mutations_per_read <- NULL
type <- "recent"

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_5mi_6s_0mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_5mi_6s_0mpr[["processed_hist"]]
```


```{r 16rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_5mi_6s_0mpr[["matrices"]])) {
  table_name <- names(rna_3mr_5mi_6s_0mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_0mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_5mi_6s_0mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_5mi_6s_0mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_0mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_5mi_6s_0mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_5mi_6s_0mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_0mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_0mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_5mi_6s_0mpr)
```

# RNA samples

## 3 min reads, 5 min tags, 6 sequencer, 10 MPR, 22<=pos<=185

```{r 17rna_tenmpr}
max_mutations_per_read <- 10

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_5mi_6s_10mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_5mi_6s_10mpr[["processed_hist"]]
```

```{r 18rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_5mi_6s_10mpr[["matrices"]])) {
  table_name <- names(rna_3mr_5mi_6s_10mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_10mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_5mi_6s_10mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_5mi_6s_10mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_10mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_5mi_6s_10mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_5mi_6s_10mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_10mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_10mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_5mi_6s_10mpr)
```

# RNA samples

## 3 min reads, 5 min tags, 6 sequencer, 5 MPR, 22<=pos<=185

```{r 20rna_givesdf}
max_mutations_per_read <- 5

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_3mr_5mi_6s_5mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_3mr_5mi_6s_5mpr[["processed_hist"]]
```


```{r 21rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_3mr_5mi_6s_5mpr[["matrices"]])) {
  table_name <- names(rna_3mr_5mi_6s_5mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_5mpr[["matrices"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_3mr_5mi_6s_5mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_3mr_5mi_6s_5mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_5mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_3mr_5mi_6s_5mpr[["matrices_counts"]])) {
  table_name <- names(rna_3mr_5mi_6s_5mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_3mr_5mi_6s_5mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_3mr_5mi_6s_5mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_3mr_5mi_6s_5mpr)
```

# RNA samples

## 5 min reads, 5 min tags, 6 sequencer, 5 MPR, 22<=pos<=185

```{r 23rna_givesdf}
min_reads <- 5

excel <- glue("excel/{rundate}_{type}_{min_reads}minread_{min_tags}mtags_{min_sequencer}minseq_{min_position}minpos_{max_position}maxpos")
if (!is.null(max_mutations_per_read)) {
  excel <- glue("{excel}_{max_mutations_per_read}mpr")
}
if (isTRUE(prune_n)) {
  excel <- glue("{excel}_prune")
}
excel <- glue("{excel}-v{ver}.xlsx")

rna_5mr_5mi_6s_5mpr <- create_matrices(
    sample_sheet=sample_sheet,
    ident_column=ident_column,
    mut_column=mut_column,
    min_reads=min_reads,
    min_tags=min_tags,
    min_sequencer=min_sequencer,
    max_mutations_per_read=max_mutations_per_read,
    min_position=min_position,
    max_position=max_position,
    prune_n=prune_n,
    verbose=verbose,
    excel=excel)

rna_5mr_5mi_6s_5mpr[["processed_hist"]]
```


```{r 24rna_mutation_tag_count, results='asis'}
## Raw matrices
for (t in 1:length(rna_5mr_5mi_6s_5mpr[["matrices"]])) {
  table_name <- names(rna_5mr_5mi_6s_5mpr[["matrices"]])[t]
  message("Raw table: ", table_name, ".")
  print(knitr::kable(rna_5mr_5mi_6s_5mpr[["matrices"]][table_name]))
  if (!is.null(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])) {
    plot(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices"]][[table_name]])
  }
}

## Matrices normalized by CPM length
for (t in 1:length(rna_5mr_5mi_6s_5mpr[["matrices_cpmlength"]])) {
  table_name <- names(rna_5mr_5mi_6s_5mpr[["matrices_cpmlength"]])[t]
  message("CPM length table: ", table_name, ".")
  print(knitr::kable(rna_5mr_5mi_6s_5mpr[["matrices_cpmlength"]][table_name]))
  if (!is.null(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])) {
    plot(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices_cpmlength"]][[table_name]])
  }
}
## Matrices normalized by counts length
for (t in 1:length(rna_5mr_5mi_6s_5mpr[["matrices_counts"]])) {
  table_name <- names(rna_5mr_5mi_6s_5mpr[["matrices_counts"]])[t]
  message("Counts table: ", table_name, ".")
  print(knitr::kable(rna_5mr_5mi_6s_5mpr[["matrices_counts"]][table_name]))
  if (!is.null(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices_counts"]][[table_name]])) {
    plot(rna_5mr_5mi_6s_5mpr[["plots"]][["matrices_counts"]][[t]])
  }
}

rm(rna_5mr_5mi_6s_5mpr)
```

# Questions from Dr. DeStefano

I think what is best is to get the number of recovered mutations of each type
from each data set.  That would be A to T, A to G, A to C; T to A, T to G, T to
C; G to A, G to C, G to T; and C to A, C to G, C to T; as well as deletions and
insertions.  I would then need the sum number of the reads that met all our
criteria (i.e. at least 3 good recovered reads for that 14 nt tag).  Each set
of 3 or more would ct as "1" read of that particular tag so I would need the
total with this in mind.  I also need to know the total number of nucleotides
that were in the region we decided to consider in the analysis.  We may want to
try this for 3 or more and 5 or more recovered tags if it is not hard.  This
information does not include specific positions on the template where errors
occurred but we can look at that latter.  Right now I just want to get a general
error rate and type of error.  It would basically be calculated by dividing the
number of recovered mutations of a particular type by sum number of the reads
times the number of nucleotides screened in the template.  As it ends up, this
number does not really have a lot of meaning but it can be used to calculate the
overall mutation rate as well as the rate for transversions, transitions, and
deletions and insertions.

# Answers

In order to address those queries, I invoked create_matrices() with a minimum
tag count of 3 and 5.  It should be noted that this is not the same as
requiring 3 or 5 reads per tag.  In both cases I require 3 reads per tag.

## Recovered mutations of each type

I am interpreting this question as the number of tags recovered for each
mutation type.  I collect this information in 2 ways of interest: the tags by
type which are deemed to be from the RT and from the sequencer.  In addition, I
calculate a normalized (cpm) version of this information which may be used to look for
changes across samples.

### Mutations by RT tag

This following block should print out tables of the numbers of mutant tags
observed for each type for the RT and the sequencer.  One would hope that the
sequencer will be consistent for all samples, but I think the results will
instead suggest that my metric is not yet stringent enough.


```{r saveme}
pander::pander(sessionInfo())
message(paste0("This is hpgltools commit: ", get_git_commit()))
this_save <- paste0(gsub(pattern="\\.Rmd", replace="", x=rmd_file), "-v", ver, ".rda.xz")
message(paste0("Saving to ", this_save))
tmp <- sm(saveme(filename=this_save))
```

```{r loadme, eval=FALSE
loadme(filename=this_save)
```
